core/src/main/scala/akka/persistence/r2dbc/journal/JournalDao.scala

```diff
@@ -242,58 +249,99 @@ private[r2dbc] class JournalDao(journalSettings: R2dbcSettings, connectionFactor
-        val slice = persistenceExt.sliceForPersistenceId(persistenceId)
-
-        val deleteMarkerSeqNrFut =
-          if (toSequenceNr == Long.MaxValue)
-            readHighestSequenceNr(persistenceId, 0L)
-          else
-            Future.successful(toSequenceNr)
-
-        deleteMarkerSeqNrFut.flatMap { deleteMarkerSeqNr =>
-          def bindDeleteMarker(stmt: Statement): Statement = {
-            stmt
-              .bind(0, slice)
-              .bind(1, entityType)
-              .bind(2, persistenceId)
-              .bind(3, deleteMarkerSeqNr)
-              .bind(4, "")
-              .bind(5, "")
-              .bind(6, 0)
-              .bind(7, "")
-              .bind(8, Array.emptyByteArray)
-              .bind(9, true)
-          }
-
-          r2dbcExecutor
-            .update(s"delete [$persistenceId]") { connection =>
-              Vector(
-                connection
-                  .createStatement(deleteEventsSql)
-                  .bind(0, persistenceId)
-                  .bind(1, toSequenceNr),
-                bindDeleteMarker(connection.createStatement(insertDeleteMarkerSql)))
-            }
-            .map(_.head)
-        }
+    val batchSize = journalSettings.cleanupSettings.eventsJournalDeleteBatchSize
+
+    def deleteInBatches(from: Long, maxTo: Long): Future[Unit] = {
+      if (from + batchSize > maxTo) {
+        deleteBatch(from, maxTo)
+      } else {
+        val to = from + batchSize - 1
+        deleteBatch(from, to).flatMap(_ => deleteInBatches(to + 1, maxTo))
       }
+    }
 
-    if (log.isDebugEnabled)
-      result.foreach(deletedRows => log.debug("Deleted [{}] events for persistenceId [{}]", deletedRows, persistenceId))
+    def insertDeleteMarker(deleteMarkerSeqNr: Long): Future[Long] = {
+      val entityType = PersistenceId.extractEntityType(persistenceId)
+      val slice = persistenceExt.sliceForPersistenceId(persistenceId)
+      r2dbcExecutor
+        .updateOne(s"insert delete marker for [$persistenceId], seq num [$deleteMarkerSeqNr]") { connection =>
+          connection
+            .createStatement(insertDeleteMarkerSql)
+            .bind(0, slice)
+            .bind(1, entityType)
+            .bind(2, persistenceId)
+            .bind(3, deleteMarkerSeqNr)
+            .bind(4, "")
+            .bind(5, "")
+            .bind(6, 0)
+            .bind(7, "")
+            .bind(8, Array.emptyByteArray)
+            .bind(9, true)
+        }
+    }
 
-    result.map(_ => ())(ExecutionContexts.parasitic)
+    if (resetSequenceNumber) {
+      for {
+        fromSeqNr <- readLowestSequenceNr(persistenceId)
+        toSeqNr <- highestSeqNr(persistenceId, toSequenceNr)
+        _ <- deleteInBatches(fromSeqNr, toSeqNr)
+      } yield ()
+    } else {
+      for {
+        fromSeqNr <- readLowestSequenceNr(persistenceId)
+        toSeqNr <- highestSeqNr(persistenceId, toSequenceNr)
+        _ <- deleteInBatches(fromSeqNr, toSeqNr)
+        _ <- insertDeleteMarker(toSeqNr)
```

Comment:
This is  a problem. The delete marker must be in the same transaction as the last delete batch. Otherwise the delete batch might succeed but not the delete marker and we end up with all events deleted but no delete marker.

Reply:
fixed